Documentation/proposals/202305-scrapeclasses.md

```diff
@@ -0,0 +1,189 @@
+
+### Pitfalls of the current solution
+
+The only known solution for use cases where you'd need to use unsafe TLS settings is to use `additionalScrapeConfig`. 
+The downside is obviously the loss of a great feature of the Prometheus Operator. The monitor resources make it possible 
+to compose the scrape configurations in a Kubernetes way.
+
+## Goals
+
+- Allow for the administrator to define a named, reusable scrape configuration snippet, including unsafe elements such as file references.
+- Allow for a user to select a configuration snippet by name in their probe/podmonitor/servicemonitor endpoint configuration.
+- Avoid giving the user the ability to use arbirary files within the Prometheus pod.
+
+### Audience
+
+- Users who serve Prometheus as a service and want to give their customers autonomy in defining monitors.
+- Users who want to scrape pods within an Istio mesh.
+
+## How
+
+The proposed solution is to introduce a notion of a  _scrape class_, akin to a Kubernetes [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/). 
+A scrape class defines TLS settings (and possibly other settings in future) to be applied to all scrape endpoints of that class. 
+
+When defining a probe/podmonitor/servicemonitor, a user may optionally assign a class to each endpoint. 
+A default class may be defined by the administrator.
+
+Class names are assumed to be installation-specific. In practice, some common class names like `istio-mtls` are likely to emerge.
+
+### Prometheus Resource
+It is proposed that the `Prometheus` resource contain a new section for defining scrape classes.
+
+_The rationale for defining scrape classes inline is that, in practice, the TLS file paths are closely related to the `volumeMounts`
+of the `Prometheus` spec. An alternative is outlined later, of factoring the class definitions into a separate resource._
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: Prometheus
+spec:
+  # optional: set a default scrape class
+  defaultScrapeClass: istio-mtls
+
+  # define scrape classes for use by the monitors
+  scrapeClasses:
+    - name: istio-mtls
+      tlsConfig:
+        caFile: "/etc/istio-certs/root-cert.pem"
+        certFile: "/etc/istio-certs/cert-chain.pem"
+        keyFile: "/etc/istio-certs/key.pem"
+        insecureSkipVerify: true
+
+  # mount the certs from the istio sidecar (shown here for illustration purposes)
+  volumeMounts:
+    - name: istio-certs
+      mountPath: "/etc/istio-certs/"
+```
+
+### PodMonitor Resource
+Allow the user to select a scrape class for each endpoint.
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: PodMonitor
+spec:
+  podMetricsEndpoints:
+  - port: http
+    path: /metrics
+    scrapeClass: istio-mtls
+```
+
+The suggested behavior is:
+1. the `tlsConfig` in the associated scrape class is automatically applied to the scrape configuration of the endpoint.
+2. the inline `tlsConfig` (if any) takes precedence over the `tlsConfig` in the scrape class.
```

Comment:
Is the resulting TLS config 1) a merge between  the scrape class and the pod monitor TLS configs or 2) only the TLS config from the pod monitor?

Reply:
Can you think of an example where a merge would be meaningful?

I have the impression that `tlsConfig` in the monitor would provide application-specific TLS settings, whereas the scrape class would provide infra-level TLS (e.g. for the Istio case where TLS is transparent).

Reply:
One example I can think of: Prometheus defines the "default" client TLS certificate and key but the monitor defines the serverName value.